util/utils.py

```diff
@@ -166,3 +169,174 @@ def get_seed_from_member_id(member_id: int) -> int:
+                lambda k=None: f"{{{k}}}",
+                {
+                    "fof_type": fof_val,
+                    "member_id": "{member_id}",
+                    "member_type": "{member_type}",
+                },
+            )
+            formatted = [item.format_map(safe_dict) for item in items]
+            expanded.append(formatted[0] if len(formatted) == 1 else tuple(formatted))
+
+    return expanded
+
+
+def expand_members(zipped, member_ids=None, member_type=None):
+    """
+    Expands a list of tuples (zip) by replacing the placeholders
+    {member_id} and {member_type} with the values provided.
+    If a placeholder is present but no corresponding value is given,
+    the placeholder is left unchanged.
+    """
+
+    member_list = to_list(member_ids)
+    member_type_list = to_list(member_type)
+    zipped = normalize_zipped(zipped)
+
+    expanded = []
+
+    for items in zipped:
+        try:
+            file_type = (FileInfo(items[0])).type
+        except (TypeError, ValueError):
+            file_type = None
+
+        placeholders = {
+            key: any(f"{{{key}}}" in str(item) for item in items)
+            for key in ("member_id", "member_type")
+        }
+
+        member_values = value_list("member_id", to_list(member_ids), placeholders)
+
+        if file_type is FileType.FOF:
+            if member_type_list:
+                member_type_list = [
+                    f"_member_id_{mtype}_" for mtype in member_type_list
+                ]
+            elif not (member_list and all(m == "ref" for m in member_list)):
+                member_type_list = ["_member_id_"]
+            else:
+                member_type_list = []
+
+        member_values_expanded = []
+
+        if file_type is FileType.STATS and member_type_list:
+            member_values_expanded = (
+                [
+                    f"{m_type}_{m_id}"
+                    for m_type in member_type_list
+                    for m_id in member_values
+                ]
+                if member_values
+                else member_type_list.copy()
+            )
+        else:
+            member_values_expanded = member_values.copy()
+
+        for member_val in member_values_expanded:
+            formatted = [
+                item.format(
+                    member_id=member_val or "{member_id}",
+                    member_type=(
+                        member_type_list[0] if member_type_list else "{member_type}"
+                    ),
+                )
+                for item in items
+            ]
+            expanded.append(formatted[0] if len(formatted) == 1 else tuple(formatted))
+
+    return expanded
+
+
+class FileType(Enum):
+    """
+    Class that memorizes the distinction between file types
+    """
+
+    FOF = "fof"
+    STATS = "csv"
+
+
+@dataclass
+class FileInfo:
+    """
+    Class that memorize the path and the type of a file.
+    """
+
+    path: str
+    type: FileType = None
```

Comment:
type is a built in python function, that is why I would suggest to use a more descriptive way.
```suggestion
    file_type: FileType = None
```
I hope this can be refactored with one click.